src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -255,8 +256,57 @@ private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey,
         /// </summary>
         public List<ImmutableArray<CallGraphNode>> GetCallCycles()
         {
-            // ToDo: need to implement a robust algorithm for finding cycles
-            throw new NotImplementedException();
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
+                {
+                    foreach (var temp in dependencies)
+                    {
+                        var (curr, _) = temp;
```

Comment:
```suggestion
                    foreach (var curr in dependencies.Keys)
                    {
```
but this means `curr` cannot be reassigned, so the "Cycle detected" section would need to be changed also.

Reply:
Yah I initially had the tuple deconstruction in the foreach, but it also did not let me update those later. So I just went with a variable that I immediately deconstruct into the value that I actually want.

Reply:
Kinda dumb that .NET does not let me do this without a dumb looking extra line like this.

Reply:
If you factor out the code that is reassigning `curr` (which I think would be good - I made a comment about that), you can use `Keys`.

Reply:
I renamed temp to make it feel less hacky. I think the one line solution here is simpler than adding a local function.